src/kv_client/mod.rs

```diff
@@ -78,23 +78,27 @@ impl KvClient for KvRpcClient {
     }
 }
 
-fn map_errors_and_trace<Resp, RpcFuture>(
+async fn map_errors_and_trace<Resp, RpcFuture>(
     request_name: &'static str,
     fut: ::grpcio::Result<RpcFuture>,
-) -> impl Future<Output = Result<Resp>>
+) -> Result<Resp>
 where
     Compat01As03<RpcFuture>: Future<Output = std::result::Result<Resp, ::grpcio::Error>>,
     Resp: HasError + Sized + Clone + Send + 'static,
 {
     let context = tikv_stats(request_name);
 
-    // FIXME should handle the error, not unwrap.
-    Compat01As03::new(fut.unwrap())
-        .map(|r| match r {
-            Err(e) => Err(ErrorKind::Grpc(e).into()),
-            Ok(r) => Ok(r),
-        })
-        .map(move |r| context.done(r))
+    let fut = match fut {
+        Err(e) => return context.done(Err(ErrorKind::Grpc(e).into())),
+        Ok(f) => f,
+    };
```

Comment:
Could this be written as `context.done(fut.and_then(async |f| Compat01As03::new(fut).await).map_err(|e| ErrorKind::Grpc(e).into()))` or something similar?

Reply:
@nrc It seems that the closure `async |f| Compat01As03::new(fut).await` could not return a Result (it returns a Future).

I tries to simplify the code as below for code review.

```
    let res = match fut {
        Ok(f) => Compat01As03::new(f).await,
        Err(e) => Err(e),
    };

    let context = tikv_stats(request_name);
    context.done(res.map_err(|e| ErrorKind::Grpc(e).into()))
```